### PR TITLE
DFC-428 |  Add language toggle

### PR DESCRIPTION
**Pull request by SophiaDWokoma**
### What changed

- Add a new feature flag in the env file called language_toggle
- Change the build-sass command
- Create a new middleware function for the creation of 2 new local variables
- Add unit test for middleware function 
- Add the new lang toggle nunjuck component into the base file.
- Remove existing implementation for lang toggle (including unit tests)
- Import addLanguageParam function from package into Nunjucks config file.
- Add function and unit test for getCurrentUrl function which is required to set` url` parameter of the lang toggle component
- Update the setLocalVarsMiddleware function to set the` currentUrl` locally using the newly implemented getCurrentUrl function.
- Updated set-local-vars-middleware test to have get method as this is required due to the getCurrentUrl function being called as part of the setLocalVarsMiddleware function

### Important to Note

Please be aware there is a current bug with the back button , which can be viewed on the Activity History page, where if  you have changed the language, it will take you back to the previous language you were on rather than the summary page. Current behaviour of the back button aligns with the anticipated behaviour and has been outlined by the UCD team here: https://docs.google.com/document/d/1RV4JZZwZ3vaiflTI07EVEg_iIhiBhhga-mVFJx2u7Z0/edit#heading=h.qi06c57hx6b9.

### Why did it change

Add a new shared component called frontend-language-toggle

### Related links

- [DFC-428](https://govukverify.atlassian.net/browse/DFC-428 

### Environment variables or secrets

- [ ] Application configuration is up-to-date
- [ ] Documented in the README
- [ ] Added to deployment steps
- [ ] Added to local startup config

## Testing

Tested that pages are correctly being translated as user toggles between languages.

## Screenshots 
**Desktop**
Screenshot 2024-03-28 at 11 50 52

**Mobile View**
Screenshot 2024-03-28 at 11 51 23

[DFC-428]: https://govukverify.atlassian.net/browse/DFC-428?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DFC-416]: https://govukverify.atlassian.net/browse/DFC-416?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Joanneyoung01** [comment]
Hey Sophia, implementation looks great, but I think the screenshots have the legacy version of the toggle as the underlining isn't on the desktop version of the toggle. Please can you attach versions of the toggle desktop/tablet + mobile for the accounts pod to see :D

File: package.json
Comment (di-fabs) [inline package.json:11]: Tell me if I'm wrong but you're using a new css file called language-toggle.css but don't use it in any file. Where is it imported?

**Review by di-fabs — CHANGES_REQUESTED** [review]

File: package.json
Comment (SophiaDWokoma) [reply package.json:11]: omg yes, good catch thank you !  I didn't realise that I hadn't added our styling because the styling from their previous implementation was being applied.

**Comment by SophiaDWokoma** [comment]
> Hey Sophia, implementation looks great, but I think the screenshots have the legacy version of the toggle as the underlining isn't on the desktop version of the toggle. Please can you attach versions of the toggle desktop/tablet + mobile for the accounts pod to see :D

Yes , I forgot to remove their styling and add ours, so have updated and replaced with the correct screenshots

File: package.json
Comment (SophiaDWokoma) [reply package.json:11]: I have added link tag to language-toggle.css in the base file

**Review by danacotoran — COMMENTED** [review]
Hi, thanks very much for your work on this! 
I just had a few comments/questions I have left as inline comments. I hope they make sense but do let me know if you have any questions about any of them.

I also wanted to note:
- There is an error when the contact triage page (`/contact-gov-uk-one-login`) is visited. This only happens when the language toggle is present, which makes me think it must be connected (I'm not sure how, but I can try to help investigate further next week as I am not working tomorrow)
- While `hreflang` was used on the language toggle link(s) to represent the language of the resource that the links are linking to, I noticed that a `lang` attribute was not used to signify that the text is in a different language to the rest of the document – is there a reason why?
- What is the `data-journey-click="link - click:lang-select:cy"` attribute for?
- Due to the hidden text, there is a bit of repetition when you are using screen reader software to navigate the page. I only tested superficially using voiceover, but to exemplify what I mean, when I am viewing the page in Welsh, and I navigate to the English language link, VO reads: "link, English Change to English, 1 of 2". This seems a bit verbose and isn't in line with HMRC's toggle.

You might want to rebase `main` as this branch seems to be out of date!

Thanks again and let me know if you have any questions!

File: deploy/template.yaml
Comment (SophiaDWokoma) [reply deploy/template.yaml:174]: I didn't update this, I think it might have been Andrew when he was helping deploy and test my branch. But I have been informed that environment flags are team's responsibility to manage as we're not the codeowners for those repos and we need to be consistent with how we work with each team

File: package.json
Comment (SophiaDWokoma) [reply package.json:13]: no, sorry will remove

File: src/components/common/layout/base.njk
Comment (SophiaDWokoma) [reply src/components/common/layout/base.njk]: Yep good shout, I will import it into your application stylesheet

File: src/components/common/layout/base.njk
Comment (SophiaDWokoma) [reply src/components/common/layout/base.njk]: removed

**Comment by SophiaDWokoma** [comment]
> Hi, thanks very much for your work on this! I just had a few comments/questions I have left as inline comments. I hope they make sense but do let me know if you have any questions about any of them.
> 
> I also wanted to note:
> 
> * There is an error when the contact triage page (`/contact-gov-uk-one-login`) is visited. This only happens when the language toggle is present, which makes me think it must be connected (I'm not sure how, but I can try to help investigate further next week as I am not working tomorrow)
> * While `hreflang` was used on the language toggle link(s) to represent the language of the resource that the links are linking to, I noticed that a `lang` attribute was not used to signify that the text is in a different language to the rest of the document – is there a reason why?
> * What is the `data-journey-click="link - click:lang-select:cy"` attribute for?
> * Due to the hidden text, there is a bit of repetition when you are using screen reader software to navigate the page. I only tested superficially using voiceover, but to exemplify what I mean, when I am viewing the page in Welsh, and I navigate to the English language link, VO reads: "link, English Change to English, 1 of 2". This seems a bit verbose and isn't in line with HMRC's toggle.
> 
> You might want to rebase `main` as this branch seems to be out of date!
> 
> Thanks again and let me know if you have any questions!

- It would be good to see the error , when you get back as I am not able to see that page/error, unless my branch is deployed
- This is something we can raise as tech debt
- data-journey-click="link - click:lang-select:cy is the attribute we’ll use to look up for analytics tracking, and will also be added to the GA4 toggle at some point it’s in our backlog https://govukverify.atlassian.net/browse/DFC-401 
-  Accessibility testing was carried out by UCD's Derren Wilson and he’s aware of this issue 
(https://govukverify.atlassian.net/browse/DFC-381), he’s given sign off with the understanding of this
- I have rebased from main now

**Review by danacotoran — COMMENTED** [review]
Hi, thanks for fixing the issue on the contact triage page 🎉 
I had a question about the parameters on the component (in the inline comment). I hope what I am asking makes sense.

Also, just wanted to confirm whether the changes requested by @di-fabs have been made 😄 

Thanks!

**Comment by SophiaDWokoma** [comment]
> Hi, thanks for fixing the issue on the contact triage page 🎉 I had a question about the parameters on the component (in the inline comment). I hope what I am asking makes sense.
> 
> Also, just wanted to confirm whether the changes requested by @di-fabs have been made 😄
> 
> Thanks!

yes they have

**Comment by danacotoran** [comment]
Thanks for your all your work on this @SophiaDWokoma!
My question from my previous review still stands, however since it's a non-blocker, I would be happy to approve this once the commit history has been tidied up and the branch has been updated with the `main` one!

**Review by danacotoran — APPROVED** [review]

**Comment by danacotoran** [comment]
@SophiaDWokoma thanks again for your work on this one. Happy for it to go in however it does seem as if it needs a ✅  from @di-fabs who has previously requested changes.

**Review by di-fabs — APPROVED** [review]

**Pull request merged**
